src/backend/distributed/planner/multi_explain.c

```diff
@@ -99,10 +182,682 @@ MultiExplainOneQuery(Query *query, IntoClause *into, ExplainState *es,
+{
+	StringInfo explainQuery = makeStringInfo();
+	char *formatStr = NULL;
+
+	switch (es->format)
+	{
+		case EXPLAIN_FORMAT_XML:
+		{
+			formatStr = "XML";
+		}
+		break;
+
+		case EXPLAIN_FORMAT_JSON:
+		{
+			formatStr = "JSON";
+		}
+		break;
+
+		case EXPLAIN_FORMAT_YAML:
+		{
+			formatStr = "YAML";
+		}
+		break;
+
+		default:
+		{
+			formatStr = "TEXT";
+		}
+		break;
+	}
+
+	appendStringInfo(explainQuery,
+					 "EXPLAIN (ANALYZE %s, VERBOSE %s, "
+					 "COSTS %s, BUFFERS %s, TIMING %s, "
+					 "FORMAT %s) %s",
+					 es->analyze ? "TRUE" : "FALSE",
+					 es->verbose ? "TRUE" : "FALSE",
+					 es->costs ? "TRUE" : "FALSE",
+					 es->buffers ? "TRUE" : "FALSE",
+					 es->timing ? "TRUE" : "FALSE",
+					 formatStr,
+					 queryString);
+
+	return explainQuery;
+}
+
+
+/* below are private functions copied from explain.c */
+
+/*
+ * Open a group of related objects.
+ *
+ * objtype is the type of the group object, labelname is its label within
+ * a containing object (if any).
+ *
+ * If labeled is true, the group members will be labeled properties,
+ * while if it's false, they'll be unlabeled objects.
+ */
+static void
+ExplainOpenGroup(const char *objtype, const char *labelname,
+				 bool labeled, ExplainState *es)
+{
+	switch (es->format)
+	{
+		case EXPLAIN_FORMAT_TEXT:
+		{
+			/* nothing to do */
+		}
+		break;
+
+		case EXPLAIN_FORMAT_XML:
+		{
+			ExplainXMLTag(objtype, X_OPENING, es);
+			es->indent++;
+		}
+		break;
+
+		case EXPLAIN_FORMAT_JSON:
+		{
+			ExplainJSONLineEnding(es);
+			appendStringInfoSpaces(es->str, 2 * es->indent);
+			if (labelname)
+			{
+				escape_json(es->str, labelname);
+				appendStringInfoString(es->str, ": ");
+			}
+			appendStringInfoChar(es->str, labeled ? '{' : '[');
+
+			/*
+			 * In JSON format, the grouping_stack is an integer list.  0 means
+			 * we've emitted nothing at this grouping level, 1 means we've
+			 * emitted something (and so the next item needs a comma). See
+			 * ExplainJSONLineEnding().
+			 */
+			es->grouping_stack = lcons_int(0, es->grouping_stack);
+			es->indent++;
+		}
+		break;
+
+		case EXPLAIN_FORMAT_YAML:
+		{
+			/*
+			 * In YAML format, the grouping stack is an integer list.  0 means
+			 * we've emitted nothing at this grouping level AND this grouping
+			 * level is unlabelled and must be marked with "- ".  See
+			 * ExplainYAMLLineStarting().
+			 */
+			ExplainYAMLLineStarting(es);
+			if (labelname)
+			{
+				appendStringInfo(es->str, "%s: ", labelname);
+				es->grouping_stack = lcons_int(1, es->grouping_stack);
+			}
+			else
+			{
+				appendStringInfoString(es->str, "- ");
+				es->grouping_stack = lcons_int(0, es->grouping_stack);
+			}
+			es->indent++;
+		}
+		break;
+	}
+}
+
+
+/*
+ * Close a group of related objects.
+ * Parameters must match the corresponding ExplainOpenGroup call.
+ */
+static void
+ExplainCloseGroup(const char *objtype, const char *labelname,
+				  bool labeled, ExplainState *es)
+{
+	switch (es->format)
```

Comment:
These probably should have a note that saying they're copied from postgres, and then grow a indent-off tag.